blocksuite/affine/shared/src/services/link-previewer-service.ts

```diff
@@ -0,0 +1,117 @@
+import type { LinkPreviewData } from '@blocksuite/affine-model';
+import { type Container, createIdentifier } from '@blocksuite/global/di';
+import { BlockSuiteError, ErrorCode } from '@blocksuite/global/exceptions';
+import { Extension } from '@blocksuite/store';
+
+import { DEFAULT_LINK_PREVIEW_ENDPOINT } from '../consts';
+import { isAbortError } from '../utils/is-abort-error';
+
+export type LinkPreviewResponseData = {
+  url: string;
+  title?: string;
+  siteName?: string;
+  description?: string;
+  images?: string[];
+  mediaType?: string;
+  contentType?: string;
+  charset?: string;
+  videos?: string[];
+  favicons?: string[];
+};
+
+export interface LinkPreviewerExtension {
+  query: (
+    url: string,
+    signal?: AbortSignal
+  ) => Promise<Partial<LinkPreviewData>>;
+  setEndpoint: (endpoint: string) => void;
+}
+
+export const LinkPreviewerProvider = createIdentifier<LinkPreviewerExtension>(
+  'LinkPreviewerService'
+);
+
+export class LinkPreviewerService
+  extends Extension
```

Comment:
If a extension can work without any browser API, it should be a store extension.